src/controller/python/chip/native/ChipMainLoopWork_WorkSchedule.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "ChipMainLoopWork.h"
+
+#include <platform/CHIPDeviceLayer.h>
+#include <semaphore.h>
+
+namespace chip {
+namespace MainLoopWork {
+namespace {
+
+struct WorkData
+{
+    std::function<void()> callback;
+    sem_t done;
+
+    WorkData() { sem_init(&done, 0 /* shared */, 0); }
+    ~WorkData() { sem_destroy(&done); }
+    void Post() { sem_post(&done); }
+    void Wait() { sem_wait(&done); }
+};
+
+void PerformWork(intptr_t arg)
+{
+    WorkData * work = reinterpret_cast<WorkData *>(arg);
+
+    work->callback();
+    work->Post();
+}
+
+} // namespace
+
+void ExecuteInMainLoop(std::function<void()> f)
+{
+
+#if CHIP_STACK_LOCK_TRACKING_ENABLED
+    if (chip::DeviceLayer::PlatformMgr().IsChipStackLockedByCurrentThread())
+    {
+        f();
+        return;
+    }
+#endif
+
+#if defined(__APPLE__)
```

Comment:
Removed the need for this - if stack lock tracking is enabled, we are always aware when we have the lock (or in case of darwin we are running on the right queue)